### ISSUE TITLE
Serilog Destructure.ByTransforming equivalent?

### ISSUE THREAD

**Issue by greenmooseSE** [NLog/NLog#3264]
Hi,

Is there a way to transform the objects used in structured logging with NLog?
Similar to Serilog's
```csharp
            Log.Logger = new LoggerConfiguration()
                .Destructure.ByTransforming<ColorTransformedDto>(x => new {R = x.Red, G = x.Green})
```

Thanks.

**Comment by snakefoot** [comment]
Not at the moment. But I guess it should not be that difficult to add an interface for injecting a custom destructure-method into the `ObjectReflectionCache`:

https://github.com/NLog/NLog/blob/dev/src/NLog/Internal/ObjectReflectionCache.cs

The cache is currently only being used by `JsonLayout` and `XmlLayout`. Could you tell what NLog target you are planning to use?

**Comment by 304NotModified** [comment]
I think this is for the message, so this is the `ValueFormatter`?

(If so, then you need to set ConfigurationItemFactory.ValueFormatter with your custom [IValueFormatter](https://nlog-project.org/documentation/v4.6.0/html/T_NLog_IValueFormatter.htm)

**Comment by snakefoot** [comment]
@304NotModified After reading this link:

https://github.com/serilog/serilog/wiki/Structured-Data#customizing-the-stored-data

Then Serilog destructure sounds very much like NLog Precalculate. So you capture some properties from a volatile object and use those for the LogEvent. Instead of holding a reference to the volatile object that might become bad later.

I guess such a transformation-logic  would have to live on the logger, and replace the LogEvent-parameters on the fly with the transformed object.

**Comment by snakefoot** [comment]
@greenmooseSE Special extraction of values from the LogEvent-parameters are usually done with laytoutrenderers. You can create your own:

https://github.com/NLog/NLog/wiki/How-to-write-a-custom-layout-renderer

**Comment by greenmooseSE** [comment]
> Could you tell what NLog target you are planning to use?

I was just doing some comparison of the logging frameworks with regards to structural logging when noticing this feature in SeriLog (via [pluralsight](https://www.pluralsight.com/courses/modern-structured-logging-serilog-seq)).

I was just curious if NLog has similar "easy" option since the feature looked interesting since it is a one-liner to the configuration without the need of creating additional sinks/formatters/rendererse etc.

So my question has been answered so you may close this if you want, thanks :).

**Comment by snakefoot** [comment]
NLog have now also added support for objectpath in `${event-properties}`:

https://github.com/NLog/NLog/wiki/EventProperties-Layout-Renderer#objectpath

Allows one to extract single property when having logged a complex object.

**Comment by snakefoot** [comment]
Seems other users also want better control of the automatic reflection logic in NLog. See now #3610

**Comment by snakefoot** [comment]
Resolved with  #3610
